Feat/training (#485)
eclipse-tractusx/puris: Training_Preparation.md

6. Testing Our Sample Asset API
6. Create your own ShellDescriptor with that asset in Part Type lifecycle following [Industry Core KIT](https://eclipse-tractusx.github.io/docs-kits/kits/Industry%20Core%20Kit/Software%20Development%20View/Digital%20Twins%20Development%20View%20Industry%20Core%20Kit#property-specificassetids) and [Digital Twin KIT](https://eclipse-tractusx.github.io/docs-kits/kits/Digital%20Twin%20Kit/Software%20Development%20View/dt-kit-software-development-view)
7. Verify that it works with a query following same approach as in 4.

## Prerequisites

- VM with Linux like Ubuntu, 8 GB of RAM available on that VM  
- docker installed on VM (Get it [here](https://docs.docker.com/engine/install/ubuntu/) )  
- 'jq' installed on VM (Get it [here](https://jqlang.github.io/jq/download/) )  
- postman installed on VM (Get it [here](https://learning.postman.com/docs/getting-started/installation/installation-and-updates/) )  
- git installed on VM (usually pre-installed on most linux systems)  
- a code editor for simple editings in JSON ([VS Code](https://code.visualstudio.com/docs/setup/linux) recommended)

## Getting Started

> - clone the PURIS project from the FraunhoferISST github [repository](https://github.com/FraunhoferISST/puris)  
> 
> - move your shell into the folder 'puris'  
> 
> 
> - ```git switch feat/training```   
> 
> 
> - Note: The above-mentioned branch is not present, if you mistakenly cloned puris from the tractusX-repository!
> 
> 
> - move your shell into the folder 'puris/local'  
> 
> 
> - ```sh launch_infrastructure.sh```  
> 
> 
> 
> Wait for the shell-script to finish (should take 1 or 2 minutes)  
> 
> 
> 
> Now, again from the 'puris/local' folder do: 
> 
>  ```docker compose up```  
> 
> 
> 
> This should take another 1 or 2 minutes. In the meantime, please start your Postman and [import](https://learning.postman.com/docs/getting-started/importing-and-exporting/importing-data/) the 'TX-Training' collection from the 'puris/local/postman' folder 

Reviewer: ```suggestion
- open a terminal and clone the PURIS project from the FraunhoferISST GitHub [repository](https://github.com/FraunhoferISST/puris) (`git clone` command)  

- after successful cloning, move your shell into the folder 'puris' folder (`cd puris`)

- Switch the branch to the training: `git switch feat/training` 
  - _Note: The above-mentioned branch is not present if you mistakenly cloned puris from the eclipse-tractusx repository!_

- move your shell into the folder 'local' (`cd local`) 

- Launch the infrastructure with `sh launch_infrastructure.sh`

Wait for the shell script to finish (it should take around 1 or 2 minutes, depending on your hardware)

Now, again from the `puris/local` folder, run the docker-compose setup with `docker compose up`

This should take another 1 or 2 minutes. In the meantime, please start your Postman and [import](https://learning.postman.com/docs/getting-started/importing-and-exporting/importing-data/) the `TX-Training` collection from the `puris/local/postman` folder 
```

Author: Tom had already covered that part in his remarks and I rewrote most of it.

> This should take another 1 or 2 minutes. In the meantime, please start your Postman and [import](https://learning.postman.com/docs/getting-started/importing-and-exporting/importing-data/) the 'TX-Training' collection from the 'puris/local/postman' folder 

Author: The Postman-Collections is not generated or edited by the script. It is right there from the start

> This should take another 1 or 2 minutes. In the meantime, please start your Postman and [import](https://learning.postman.com/docs/getting-started/importing-and-exporting/importing-data/) the 'TX-Training' collection from the 'puris/local/postman' folder 

Reviewer 2: You're right. You fixed the keys :) resolved

> 

Reviewer 2: Lets also list relevant swagger uis:

After starting everything you can use the following links to access open APIs. Come back to them for further information needed:
- [DTR](http://dtr-supplier:4243/swagger-ui/index.html#/Asset%20Administration%20Shell%20Registry%20API/PostSubmodelDescriptor-ThroughSuperpath)
- [TX EDC](https://app.swaggerhub.com/apis/eclipse-tractusx-bot/tractusx-edc/0.7.1)
- [PURIS](http://localhost:8081/catena/swagger-ui/index.html#/demand-and-capacity-notification-request-api-controller/postDemandAndCapacityNotification) (NOT NEEDED BY TASKS, but maybe interesting)

## Your tasks

> 

Reviewer 2: How about adding a small intro like: This training is divided into two conceptual sections: Tasks 1 to 3 will support you in understanding the concepts using the concepts mainly for read and some DTR write operations. Tasks >= 4 will support guide you to create a simple submodel and make it available through the EDC + EDC combination.
In both sections you will use postman to do the tasks performed by applications of the data provider and data consumer.

> ### Task 1: Simple Interaction with the Digital Twin Registry

Reviewer: ```suggestion
### Task 1: Simple Interaction with the Digital Twin Registry (DTR)
```

Author: Done

We want to spare you the trouble of copying these data individually and transferring them to other request bodies.

If we now look at the current request body, we see that we use the string "urn:samm:io.catenax.part_type_information:1.0.0#PartTypeInformation" as a filter criterion to start another catalog query.

#### i) Negotiation for a Contract over the PartType Asset

Business as usual :)

#### j) Querying the Negotiation Status
As a response, we receive the contract ID.

#### k) Requesting a Pull Transfer
We provide the contract ID.

#### l) Requesting an EDR
As above.

#### m) Pull Data from PartType-API

We now see what information the PartType-API provides regarding this material. It's also interesting to note that the PURIS backend internally handled the response. If we take a quick look at the logs of our PURIS backend containers, we can see that the PartTypeInformationController has processed a request.

### Task 4: Creating a New Sample Asset on Our EDC

> 

Reviewer 2: short introduction needed with some metdata on material and role (we're supplier)

#### a) Asset Registration

With this request, we create a new asset on the EDC of the supplier partner.

Let's briefly look at the request body. We create a new fictitious API asset there, which we have given the identifier "urn:samm:io.catenax.sample_information:1.0.0#SampleInformation".

Since we do not have an endpoint in the PURIS backend that can serve this asset, we simply use any URL in the "baseUrl" field that returns a list of artificial user names.

#### b) Creating a Contract Definition

Here, we reference the newly created asset in the request body.

> Additionally, we reference in the "accessPolicy" and "contractPolicy" fields two policies that our PURIS backend already created on the EDC during its boot process.

Reviewer 2: Append: In case you're further interested, checkout the [PURIS documentation](https://github.com/eclipse-tractusx/puris/blob/main/docs/architecture/08_concepts.md#data-sovereignty) explaining the policies used and the data sovereignty concept.

### Task 5: Inserting the Sample Asset API into a Shell Descriptor

#### a) Request a token:
You know the drill ;)

#### b) Lookup for a specific Shell Descriptor:
As before.

#### c) Querying a Specific Shell Descriptor:

If we briefly look at the response, we will of course notice that our SampleAsset API is not included there.

> #### d) Modifying an Existing Shell Descriptor:

Reviewer 2: I think I found something interesting we overlooked: You can just add a submodel using this endpoint: http://dtr-supplier:4243/swagger-ui/index.html#/Asset%20Administration%20Shell%20Registry%20API/PostSubmodelDescriptor-ThroughSuperpath Maybe we can give this as a "primer task" to let them try it out for us? :)

Reviewer 2: fair, we don't follow that direction. Resolved + thanks for checking :)

Let's first calmly look at the request body. We have already manually inserted an entry for our SampleAsset API there. With this PUT request, we ensure that it can now be found for the material.

#### e) Querying a Specific Shell Descriptor Again

Let's briefly check if the previous operation was successful.

#### f) Task for Fast Completers:
Redo the steps from Task 3 a to g. Do you also find our SampleAsset API?

### Task 6: Testing Our Sample Asset API

> 

Reviewer 2: Add some support: As a customer we now are going to retrieve the data added on supplier side in the previous task.

### Task 1: Simple Interaction with the Digital Twin Registry (DTR)

In this step you will directly interact with the Digital Twin Registry, which is secured by an Identity Provider (IDP, Keycloak). You will query shells by given information
(see [Industry Core KIT](https://eclipse-tractusx.github.io/docs-kits/category/industry-core-kit)) and see a shell descriptor. We'll act as the Supplier. 

#### a) Request a JWT token for DTR Authentication / Authorization

We will retrieve a token from the Keycloak server.  
That token is needed to authenticate with the DTR. It will be automatically included in the subsequent requests. 
Please note that the token is only valid for 5 minutes. Therefore, it may be necessary to request a new token  
if an error message appears, that indicates an expired token. 

#### b) Display a list of all Shells

Before sending this request, please to a look into the headers: Note that we specify a field "edc-bpn" there. 
This is actually done in all subsequent requests of this task. 

After doing the request, we should see a large JSON file. It might be convenient to copy-and-paste it to your code editor. 

> When looking at the structure, you will get a first impression of a shell descriptor. 

Author: I assume that you are already doing some explanation on that issue in the "theory part" of the session. Perhaps you could share a smalll text bit from your theory presentation, that can be inserted here.

This flow is also described in the Digital Twin KIT and is used for standards 
relying on the Digital Twin KIT. Whenever you're wondering about the EDC steps, look them up in the 
[management api walkthrough of the EDC](https://github.com/eclipse-tractusx/tractusx-edc/tree/main/docs/usage/management-api-walkthrough).

#### a) Catalog Query for the DTR Asset

Please look at the request body. We use a filter there that should yield a unique result. The filter should find exactly 
the asset that the other partner has created for his DTR on his own DTR. 

#### b) Negotiation for a Contract over the Asset

The request body contains parts of the catalog entry received in the previous request. It has been copied in the 
previous step (automatically via Postman Scripting). Please revisit a) and the script. If you like to, try to build 
it on your own. Following the [Connector Kit](https://eclipse-tractusx.github.io/docs-kits/category/connector-kit), 
this policy node needs to be **exactly the same as offered by the supplier**.

#### c) Querying the Negotiation Status
As a response, we receive the contract ID.

#### d) Requesting a Pull Transfer

> We provide the contract ID.

Reviewer: Please add one or two sentences about what this step does.

#### e) Requesting an EDR

The EDR contains the data (authentication information and URL) with which we can subsequently access the asset. 
The authentication token also only has a validity duration of a few minutes. Please note that the EDR may need to be 

> renewed by re-executing this request. 

Reviewer: This is the first time people will work with EDR. Please add more details here and describe in more detail what an EDR is, why it is used, etc.

query. Both are forwarded by the EDC to the DTR. Accordingly, the response should be identical to what we saw in Task 1 c.  

By the way, do you remember the HTTP-header named "edc-bpn", which we were looking at during task 1? Is it needed here too? 

In case you are unsure, have another look at the [DTR Documentation](https://github.com/eclipse-tractusx/sldt-digital-twin-registry/tree/main/docs#access-control-to-digital-twins-based-on-the-bpn-business-partner-number-tenantid)

#### g) Querying a Specific Shell via EDC Transfer

This step corresponds to the step from Task 1 d, but using an EDC transfer. Note that we can use the same contract as 
the basis for all requests. We can even use the same transfer that we requested in Task 3 d. Only the token may need 
to be renewed, if it has expired.

#### h) Catalog Query for the PartTypeApi Asset

First, let's look again at the response with the shell descriptor from the previous query. It is recommended to copy 
and paste the content into the code editor. Then, if we search through the "submodelDescriptors" array, we find an entry 
that contains the following: "urn:samm:io.catenax.part_type_information:1.0.0#PartTypeInformation".

> Let's take a closer look at this section. Within this JSON object, there is also an "endpoints" array. The entries 

Reviewer: ```suggestion
Let's take a closer look at this section. Within this JSON object, there is also an "endpoints" array (maybe above, so you have to scroll up). The entries 
```

"href" and "subprotocolBody" are particularly interesting. Evaluating this information allows us to know where and 
how we can access the submodel described in this shell descriptor.

We want to spare you the trouble of copying these data individually and transferring them to other request bodies.

If we now look at the current request body, we see that we use the string "urn:samm:io.catenax.part_type_information:1.0.0#PartTypeInformation" 
as a filter criterion to start another catalog query.

#### i) Negotiation for a Contract over the PartType Asset

Business as usual :)

#### j) Querying the Negotiation Status

As a response, we receive the contract ID.

#### k) Requesting a Pull Transfer
We provide the contract ID.

#### l) Requesting an EDR

As above.

#### m) Pull Data from PartType-API

> And now, please have another look the response body from request 3 g. Please select and copy the URL in the "href" field 

Reviewer: ```suggestion
And now, please have another look at the response body from request 3 g. Please select and copy the URL in the "href" field 
```

of the PartType submodel section, we had found there.  

> Insert it into the URL line of Postman. But we are not done yet, since we have to append another path parameter. 

Reviewer 2: I'm loving this part. We'll done!

I.e. please add ```/$value``` as the last path parameter, right after that, what you got from "href".  

The ```$value``` in the end is important at this is the Digital Twin KITs minimum requirement. It tells the endpoint to 
provide a json serialization of the whole submodel.

> We now see what information the PartType-API provides regarding this material. It's also interesting to note that the 

Reviewer: ```suggestion
After clicking the "Send" button, we can now see the information the PartType-API provides regarding this material. It's also interesting to note that the 
```

> PURIS backend internally handled the response. If we take a quick look at the logs of our PURIS backend containers, 

Reviewer: ```suggestion
PURIS backend internally handled the response. If we take a quick look at the logs of our PURIS backend containers (reopen the running docker compose terminal and search for the line starting with `supplier-backend`), 
```

nothing else (meaning we only allow GET operations). This information is basically defined by the Digital Twin KIT - 
depending on your api endpoint, you may need to update different proxy properties individually. 

#### b) Creating a Contract Definition

Here, we reference the newly created asset in the request body.

Additionally, we reference in the "accessPolicy" and "contractPolicy" fields two policies that our PURIS backend already 
created on the Tractus-X EDC during its boot process. 

If you're interested in further details in this regard, please see the [PURIS documentation](docs/architecture/08_concepts.md) 
explaining the policies used and the data sovereignty concept. 

### Task 5: Inserting the Sample Asset API into a Shell Descriptor

In this step, we want to make our newly created asset available in a shell descriptor of the supplier's DTR. 

#### a) Request a JWT token for DTR Authentication / Authorization
You know the drill ;)

#### b) Lookup for a specific Shell Descriptor
As before.

#### c) Querying a Specific Shell Descriptor

If we briefly look at the response, we will of course notice that our SampleAsset API is not included there.

> #### d) Modifying an Existing Shell Descriptor

Reviewer: ```suggestion
#### d) Add Submodel to Existing Shell Descriptor
```

> Let's first calmly look at the exiting request body of this request. We have already manually inserted an entry 

Reviewer: We need to ensure that everyone can follow here. We should maybe add one or two sentences to explain what it is done here (that we are trying to add the submodel to the already existing and previously queried shell descriptor).

Author: Fair point. After taking a closer look, I also noticed, that the pre-defined submodel there is not quite correct. The "href" should reference the public api endpoint of the data plane (and not just the base URL of the data plane. I will fix that.

for our SampleAsset API there. With this PUT request, we will now ensure that it can now be found for the material.

#### e) Querying a Specific Shell Descriptor Again

> Let's briefly check if the previous operation was successful.

Reviewer: ```suggestion
Let's briefly check if the previous operation was successful. Check if you can find our newly added Sample.
```

#### f) Task for Fast Completers:
Redo the steps from Task 3 a to g. Do you also find our SampleAsset API?

### Task 6: Testing Our Sample Asset API

We are working from the customer's perspective here, and we will perform the necessary steps in order to obtain access 
to the new Sample Asset, that the supplier has created recently. 

> 

Reviewer 2: We need to append some note here:

Note: This task is simplified as the sample endpoint does not provide the possibility to have `$value` suffix. Additionally, The DTR is not queried first - as a Fast Completer you did the query and likely found the API made available as a Submodel Descriptor. Also in that case you would normally need to extract the information for the href - same as previously - and append the `$value`.
